### ISSUE TITLE
Img Writer: ICS metadata errors

### ISSUE THREAD

**Issue by maierl** [knime-ip/knip#397]
The ICS metadata has various errors. I am aware that it is currently not possible to write more than one channel (https://github.com/scifio/scifio/issues/326), but there are a few more things to be corrected.

- the third dimension is Z for my images, but stored as "u"
- the scaling unit of the third dimension is um, but stored as "null"
- the scaling values are wrong. imagej expects a height / width / length of each pixel in units, but the values stored are something like number of pixels
- IMHO the .ics filename should be a relative path instead.

current output:
```
	
ics_version	1.0
filename	/path/img.ics
layout	parameters	6
layout	order	bits	x	y	u	
layout	sizes	16	512	512	31	0	0	
representation	format	integer
representation	sign	unsigned
representation	compression	uncompressed
representation	byte_order	2	1	
parameter	scale	1.000000	512	512	1.0	
parameter	units	bits	um	um	null	

end
```

should be:
```
	
ics_version	1.0
filename	/path?!/img.ics
layout	parameters	6
layout	order	bits	x	y	z	
layout	sizes	16	512	512	31	0	0	
representation	format	integer
representation	sign	unsigned
representation	compression	uncompressed
representation	byte_order	2	1	
parameter	scale	1.000000		0.2000000	0.20000000	2.00000000
parameter	units	bits	um	um	um

end
```

**Comment by maierl** [comment]
beyond that, the .ics writer apparently writes only the first 3 dimensions of an image, ignoring the rest.

**Comment by dietzc** [comment]
I'm sure this is a problem in `SCIFIO` or `BioFormats`, right? Can you open the issue there please? Maybe there already exists another issue related to this problem. Can you reproduce the problem in ImageJ?

**Comment by maierl** [comment]
When I write a .ics from ImageJ, the header states the seemingly right dimensionality and scaling, but the dimensions get mixed up when opening the file. That’s why I only referenced the according issue of SCIFIO.

**Comment by dietzc** [comment]
Which KNIME version are you using? Actually, we have no specific `ICS` code on KNIME side and fully rely on `SCIFIO`. Therefore, the behaviour should exactly be the same if the versions of the libraries are the same. did you use modern mode to test SCIFIO?

**Comment by maierl** [comment]
It is 3.2 stable. I didn't find any option to write .ics directly in ImageJ, so I used the bioformats exporter. I assume that does not use SCIFIO even when modern mode is activated (or --ij2 on the windows command line in my case).
@imagejan you reported the problem with SCIFIO, so maybe you could tell us how to save the image with SCIFIO instead of bioformats?
Anyway, the errors of saving with the exporter were mostly the same for me as the one imagejan mentioned.

The bug that appears using Knime reminded me a bit of the `map` variable that caused the ilastik node to fail writing multidimensional images whose order of dimensions was different from the example workflow. That is using SCIFIO down the line, too, isn't it?

**Comment by dietzc** [comment]
@maierl we fixed several minor problems in our nightly build (as it used a new SCIFIO / bioformats). Can you check if some of the problems are gone? Sorry for the trouble!

**Comment by maierl** [comment]
with the latest nightly, the image errors look just as imagejan described. additionally, all calibrations except for Y are correctly saved. Y is reset to 1.

**Comment by dietzc** [comment]
@maierl can you post a very, very small workflow to reproduce the problem (i.e. Image Generator -> Set Image Metadata -> Img Writer or so?). This would be super helpful!

**Comment by maierl** [comment]
There you go:
[ics test.knwf.txt](https://github.com/knime-ip/knip/files/476395/ics.test.knwf.txt)

**Comment by maierl** [comment]
oh and one more thing: I didn't set offsets on the image as I didn't expect them to be an intended property of regular images. is that right?

**Comment by dietzc** [comment]
I think http://forum.imagej.net/t/release-of-bio-formats-5-2-2/2760/2 is related, correct? See post of oburri.

**Comment by maierl** [comment]
do you happen to know if the SCIFIO .ics writer still uses bioformats code?
The effects of the bioformats bug look different.

**Comment by maierl** [comment]
the updated bioformats plugin works fine. yet that doensn't immediately help reading / writing files in knime I assume?

**Comment by dietzc** [comment]
> oh and one more thing: I didn't set offsets on the image as I didn't expect them to be an intended property of regular images. is that right?

I don't understand what you mean.

> do you happen to know if the SCIFIO .ics writer still uses bioformats code?
https://github.com/scifio/scifio/blob/master/src/main/java/io/scif/formats/ICSFormat.java

> the updated bioformats plugin works fine. yet that doensn't immediately help reading / writing files in knime I assume?

We use the SCIFIO format.

**Comment by maierl** [comment]
>> I didn't set offsets on the image as I didn't expect them to be an intended property of regular images. is that right?

> I don't understand what you mean.

bitmasks typically have offsets. when I would save them as image files and load them again, would the offset be restored? I didn't expect that and thus I didn't test the offests for .ics files.

**Comment by dietzc** [comment]
any kind of interval can have an offset (=minimum of interval).

**Comment by imagejan** [comment]
@maierl I agree, the offsets that you set within KNIME will not be part of the metadata in the .ics file being written. Or is this information part of `net.imagej.axis.Axes`, @dietzc ?

**Comment by dietzc** [comment]
Sorry, I misunderstood the problem. Yes, minimum is _not_ written!

**Comment by maierl** [comment]
the new release is way better for .ics than before 👍 

image dimensions stay the same now, unless they are 1 and thrown out (a pity for dimension X. probably due to https://github.com/scifio/scifio/issues/290 ?)
a few more bugs are shown by the workflow I posted here.

in my eyes the biggest problem of the new stable release is that the image metadata fails for .ics as it does for .czi files ( #404 , probably the same reason as #302 ? )
